src/providers/RecentMessagesApi.hpp

```diff
@@ -0,0 +1,247 @@
+
+            auto message =
+                Communi::IrcMessage::fromData(content.toUtf8(), nullptr);
+
+            if (message->command() == "CLEARCHAT")
+            {
+                message = convertClearchatToNotice(message);
+            }
+
+            messages.emplace_back(std::move(message));
+        }
+
+        return messages;
+    }
+
+    // Build Communi messages retrieved from the recent messages API into
+    // proper chatterino messages.
+    std::vector<MessagePtr> buildRecentMessages(
+        std::vector<Communi::IrcMessage *> &messages, Channel *channel)
+    {
+        auto &handler = IrcMessageHandler::instance();
+        std::vector<MessagePtr> allBuiltMessages;
+
+        for (auto message : messages)
+        {
+            if (message->tags().contains("rm-received-ts"))
+            {
+                QDate msgDate =
+                    QDateTime::fromMSecsSinceEpoch(
+                        message->tags().value("rm-received-ts").toLongLong())
+                        .date();
+
+                // Check if we need to insert a message stating that a new day began
+                if (msgDate != channel->lastDate_)
+                {
+                    channel->lastDate_ = msgDate;
+                    auto msg = makeSystemMessage(
+                        QLocale().toString(msgDate, QLocale::LongFormat),
+                        QTime(0, 0));
+                    msg->flags.set(MessageFlag::RecentMessage);
+                    allBuiltMessages.emplace_back(msg);
+                }
+            }
+
+            auto builtMessages = handler.parseMessageWithReply(
+                channel, message, allBuiltMessages);
+
+            for (auto builtMessage : builtMessages)
+            {
+                builtMessage->flags.set(MessageFlag::RecentMessage);
+                allBuiltMessages.emplace_back(builtMessage);
+            }
+        }
+
+        return allBuiltMessages;
+    }
+
+    // Returns the URL to be used for querying the Recent Messages API for the
+    // given channel.
+    QUrl constructRecentMessagesUrl(const QString &name)
+    {
+        QUrl url(Env::get().recentMessagesApiUrl.arg(name));
+        QUrlQuery urlQuery(url);
+        if (!urlQuery.hasQueryItem("limit"))
+        {
+            urlQuery.addQueryItem(
+                "limit",
+                QString::number(getSettings()->twitchMessageHistoryLimit));
+        }
+        url.setQuery(urlQuery);
+        return url;
+    }
+
+}  // namespace
+
+class RecentMessagesApi
+{
+public:
+    /**
+     * @brief Loads recent messages for a channel using the Recent Messages API
+     * 
+     * @param channelName Name of Twitch channel
+     * @param channelPtr Weak pointer to Channel to use to build messages
+     * @param onLoaded Callback taking the built messages as a std::vector<MessagePtr>
+     * @param onError Callback called when the network request fails
+     */
+    template <typename OnLoaded, typename OnError>
```

Comment:
Since the callbacks are well-defined in their parameters and return value, I'd prefer if they were defined as `std::function`'s in the parameters.
That would also allow us to make the file a standard header/source file combo.

Reply:
Updated this. Was also thinking of doing this because it would probably have needed to be done anyways when #3901 is updated with this PR.